src/mcpwm.rs

```diff
@@ -0,0 +1,743 @@
+//! ```
+//! let peripherals = Peripherals::take().unwrap();
+//! let config = OperatorConfig::default().frequency(25.kHz().into());
+//! let mcpwm = Mcpwm::new(peripherals.mcpwm0.mcpwm)?;
+//! let mut operator = Operator::new(
+//!     peripherals.mcpwm0.operator0,
+//!     &mcpwm,
+//!     &config,
+//!     peripherals.pins.gpio4,
+//!     peripherals.pins.gpio5,
+//! )?;
+//!
+//! println!("Starting duty-cycle loop");
+//!
+//! for &duty in [0.0, 20.0, 40.0, 60.0, 80.0, 100.0].iter() {
+//!     println!("Duty {}%", duty);
+//!     operator.set_duty_a(duty)?;
+//!     operator.set_duty_b(100.0 - duty)?;
+//!     FreeRtos.delay_ms(2000)?;
+//! }
+//! ```
+//!
+//! See the `examples/` folder of this repository for more.
+
+use core::borrow::Borrow;
+
+use crate::gpio::OutputPin;
+use crate::units::{FromValueType, Hertz};
+use esp_idf_sys::*;
+
+// MCPWM clock source frequency for ESP32 and ESP32-s3
+const MCPWM_CLOCK_SOURCE_FREQUENCY: u32 = 160_000_000;
+
+// Max PWM timer prescaler
+const MAX_PWM_TIMER_PRESCALE: u32 = 0x1_00;
+
+// Max PWM timer period
+const MAX_PWM_TIMER_PERIOD: u32 = 0x1_00_00;
+
+/// The Motor Control Pulse Width Modulator peripheral
+pub struct Peripheral<U: Unit> {
+    pub mcpwm: MCPWM<U>,
+    pub operator0: OPERATOR0<U>,
+    pub operator1: OPERATOR1<U>,
+    pub operator2: OPERATOR2<U>,
+}
+
+impl<U: Unit> Peripheral<U> {
+    /// # Safety
+    ///
+    /// It is safe to instantiate this exactly one time per `Unit`.
+    pub unsafe fn new() -> Self {
+        Self {
+            mcpwm: MCPWM::new(),
+            operator0: OPERATOR0::new(),
+            operator1: OPERATOR1::new(),
+            operator2: OPERATOR2::new(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DutyMode {
+    ActiveHigh,
+    ActiveLow,
+}
+
+impl From<DutyMode> for mcpwm_duty_type_t {
+    fn from(val: DutyMode) -> Self {
+        match val {
+            DutyMode::ActiveHigh => mcpwm_duty_type_t_MCPWM_DUTY_MODE_0,
+            DutyMode::ActiveLow => mcpwm_duty_type_t_MCPWM_DUTY_MODE_1,
+        }
+    }
+}
+
+// TODO: For UpDown, frequency is half of MCPWM frequency set
+#[derive(Clone, Copy, Debug)]
+pub enum CounterMode {
+    #[cfg(not(esp_idf_version = "4.3"))]
+    Frozen,
+    Up,
+    Down,
+    UpDown,
+}
+
+impl From<CounterMode> for mcpwm_counter_type_t {
+    fn from(val: CounterMode) -> Self {
+        match val {
+            #[cfg(not(esp_idf_version = "4.3"))]
+            CounterMode::Frozen => mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
+            CounterMode::Up => mcpwm_counter_type_t_MCPWM_UP_COUNTER,
+            CounterMode::Down => mcpwm_counter_type_t_MCPWM_DOWN_COUNTER,
+            CounterMode::UpDown => mcpwm_counter_type_t_MCPWM_UP_DOWN_COUNTER,
+        }
+    }
+}
+
+// TODO: Note that `red` and `fed` from the IDF's perspecitve is time as in number of clock cycles after the
+//       MCPWM modules group prescaler. How do we want to expose this? Do we expose it as just that, a cycle count?
```

Comment:
Raw cycles are fine for an initial run. You could try using types in `units.rs` but they not be exactly what we need.